dan_layer/core/src/workers/hotstuff_waiter.rs

```diff
@@ -852,27 +929,63 @@ where
         tx: &mut TShardStore::Transaction<'_>,
     ) -> Result<(), HotStuffError> {
         let shard = node.shard();
-        if tx.get_last_executed_height(shard)? < node.height() {
-            info!(
+        info!(
+            target: LOG_TARGET,
+            "🔥 OnCommit for payload {} {} and shard {}",
+            node.payload_height(),
+            node.payload_id(),
+            shard,
+        );
+
+        if *node.parent() != TreeNodeHash::zero() {
+            let parent = tx.get_node(node.parent())?;
+            let changes = extract_changes_from_justify(parent.justify())?;
+            self.on_commit(&parent, &changes, tx)?;
+        }
+
+        if node.justify().payload_height() == self.consensus_constants.max_payload_height() - NodeHeight(1) {
+            assert_eq!(*node.justify().decision(), QuorumDecision::Accept);
+            tx.save_substate_changes(changes, node)?;
+        }
+
+        // TODO: We shouldn't ignore the "no pledges found" error but it happens and breaks things so need to
+        //       investigate
+
+        error!(
+            target: LOG_TARGET,
+            "COMPLETE PLEDGe FOR shard {} pid {} h: {}",
+            node.shard(),
+            node.payload_id(),
+            node.payload_height()
+        );
+        if let Err(err) = tx.complete_pledges(node.shard(), node.payload_id(), node.hash()) {
+            error!(
                 target: LOG_TARGET,
-                "🔥 OnCommit for payload {} and shard {}",
-                node.payload_id(),
-                shard,
+                "Failed to complete pledges {} {}: {}",
+                node.payload_height(),
+                node.justify().payload_height(),
+                err
             );
+        }
 
-            if *node.parent() != TreeNodeHash::zero() {
-                let parent = tx.get_node(node.parent())?;
-                let changes = extract_changes_from_justify(parent.justify())?;
-                self.on_commit(&parent, &changes, tx)?;
-            }
+        Ok(())
+    }
 
-            if node.justify().payload_height() == NodeHeight(self.consensus_constants.hotstuff_rounds - 2) &&
-                *node.justify().decision() == QuorumDecision::Accept
-            {
-                tx.save_substate_changes(changes, node)?;
-                tx.complete_pledges(node.shard(), node.payload_id(), node.hash())?;
-            }
-            tx.set_last_executed_height(shard, node.height())?;
+    fn on_reject(
+        &self,
+        node: &HotStuffTreeNode<TAddr, TPayload>,
+        tx: &mut TShardStore::Transaction<'_>,
+    ) -> Result<(), HotStuffError> {
+        if *node.parent() != TreeNodeHash::zero() {
```

Comment:
You actually need to check the results of each QC here and commit or reject depending on the QC. This step of recursively applying the result to the parent is for nodes that have missed consensus rounds and are catching up. It might never really happen in Cerberus, but it's part of the original Hotstuff implementation